README.md
# n9-node-utils

Neo9 Node Utils Module.

[![npm version](https://img.shields.io/npm/v/@neo9/n9-node-utils.svg)](https://www.npmjs.com/package/@neo9/n9-node-utils)
[![Travis](https://img.shields.io/travis/neo9/n9-node-utils/master.svg)](https://travis-ci.org/neo9/n9-node-utils)
[![Coverage](https://img.shields.io/codecov/c/github/neo9/n9-node-utils/master.svg)](https://codecov.io/gh/neo9/n9-node-utils)
[![license](https://img.shields.io/github/license/neo9/n9-node-utils.svg)](https://github.com/neo9/n9-node-utils/blob/master/LICENSE)

## Installation

```bash
npm install --save @neo9/n9-node-utils
```

or

```bash
yarn add --dev @neo9/n9-node-utils
```

### V2 Breaking changes

- Drop Node 8 support
- Rename `N9JsonStreamResponse` to `N9JSONStreamResponse`

## Contents

- [N9Error](#n9error)
- [ok](#ok)
- [cb](#cb)
- [waitFor](#waitfor)
- [waitForEvent](#waitforevent)
- [asyncObject](#asyncobject)

## N9Error

Custom error class (extends `Error`), used by [n9-node-micro](https://github.com/neo9/n9-node-micro) for API errors with status code and context.

```ts
new N9Error(message [, statusCode] [, context])
```

Arguments:

- message: `String`, **required**
- status: `Number`, default: `500`
- context: `Object`, default: `{}`

Example:

```ts
import { N9Error } from '@neo9/n9-node-utils';

throw new N9Error('file-not-found', 404, { path: '/tmp/my-file.txt' });
```

## ok

Waits for the value of promise. If promise throws an Error, returns `undefined`.

```ts
ok(promise: Object): Promise
```

Arguments:

- promise: `Promise`

Example:

```ts
import { readFile } from 'fs-extra';
import { ok } from '@neo9/n9-node-utils';

// readFile sends back a Promise since we use fs-extra
const file = await ok(readFile('./my-file.txt', 'utf-8'));

if (file) console.log('File found:', file);
```

## cb

Calls a function `fn` that takes arguments `args` and an `(err, result)` callback. Waits for the `callback` result, throwing an `Error` if `err` is truthy.

```ts
cb(fn: Function, ...args: any[]): Promise
```

Arguments:

- fn: `Function`, a function that takes a callback
- args: (...`any`) arguments to pass to `fn`

Example:

```ts
import { cb } from '@neo9/n9-node-utils';

const file = await cb(readFile('./my-file.txt', 'utf-8'));

console.log('File content:', file);
```

## waitFor

Waits for ms milliseconds to pass, use `setTimeout` under the hood.

```ts
waitFor(ms: number): Promise
```

Arguments:

- ms: `Number`, default: `0`

Example:

```ts
import { waitFor } from '@neo9/n9-node-utils';

await waitFor(1000); // wait for 1s
```

## waitForEvent

Waits for emitter to emit an eventName event.

```ts
waitForEvent(emitter: EventEmitter, eventName: string): Promise<Array>
```

Arguments:

- emitter: `EventEmitter`
- eventName: `String`

Example:

```ts
import { waitForEvent } from '@neo9/n9-node-utils';

await waitForEvent(sever, 'listen');
```

## asyncObject

Waits for all Promises in the keys of `obj` to resolve.

```ts
asyncObject(obj: Object): Promise<Object>
```

Arguments:

- obj: `Object`, default: `{}`

Example:

```ts
import { asyncObject } from '@neo9/n9-node-utils';

const results = await asyncObject({
	pictures: getPictures(),
	comments: getComments(),
	tweets: getTweets(),
});

console.log(results.pictures, results.comments, results.tweets);
```


CHANGELOG.md


# [3.0.0](https://github.com/neo9/n9-node-utils/compare/3.0.0-rc.0...3.0.0) (2023-11-14)


### github

* Add conditional lcov file generation ([e57576d](https://github.com/neo9/n9-node-utils/commit/e57576d8d16480f2f0d71abf5f45e70c641fb40c))

### yarn

* Fix husky dependency declaration ([7bf3cd9](https://github.com/neo9/n9-node-utils/commit/7bf3cd9615635bd3676a94432e8476553e342a21))
* Upgrade dependencies and drop node 16 support in tests ([9b5f01c](https://github.com/neo9/n9-node-utils/commit/9b5f01c45a30f6ef736897144ea9b517eb5c6428))

# [3.0.0-rc.0](https://github.com/neo9/n9-node-utils/compare/2.2.1...3.0.0-rc.0) (2023-10-13)


### node

* Drop support of node below 16 and support for node 20 #30 ([0e6f807](https://github.com/neo9/n9-node-utils/commit/0e6f807165beca85a932b2001189204f7b49d23f)), closes [#30](https://github.com/neo9/n9-node-utils/issues/30)

### yarn

* Upgrade codecov sending ([000261f](https://github.com/neo9/n9-node-utils/commit/000261fe9409c0dcc6a9438b46a3dd84657497c3))
* Upgrade dependencies #31 ([02a66d1](https://github.com/neo9/n9-node-utils/commit/02a66d16377a69d11e983ab8bb81316de1bb9d48)), closes [#31](https://github.com/neo9/n9-node-utils/issues/31)

## [2.2.1](https://github.com/neo9/n9-node-utils/compare/2.2.0...2.2.1) (2022-07-26)


### types

* Fix type metadata usage on n9-json-stream ([3fdb85a](https://github.com/neo9/n9-node-utils/commit/3fdb85a1673d2049f419fb77a45859709edeb5ab))

# [2.2.0](https://github.com/neo9/n9-node-utils/compare/2.1.0...2.2.0) (2022-07-26)


### build

* Use github-actions for build instead of travis ([b7800bf](https://github.com/neo9/n9-node-utils/commit/b7800bfbf3099d0f97f730e64f71abad646073a2))

### types

* Add type on n9-json-stream metadata ([21e000c](https://github.com/neo9/n9-node-utils/commit/21e000c89db1a62b23a08d0ea22ccdcc1dbd145a))

# [2.1.0](https://github.com/neo9/n9-node-utils/compare/2.0.2...2.1.0) (2021-05-18)


### node

* Drop support of node 10 and add node 16 ([a31ef82](https://github.com/neo9/n9-node-utils/commit/a31ef82638000c88433690eadbffe93789087b45))

### yarn

* Update all dev dependencies ([c3df3f9](https://github.com/neo9/n9-node-utils/commit/c3df3f9320d587844495211c9acb6dcef8833cc1))

## [2.0.2](https://github.com/neo9/n9-node-utils/compare/2.0.1...2.0.2) (2020-10-12)


### n9errors

* Remove readonly on context ([a5dc9ac](https://github.com/neo9/n9-node-utils/commit/a5dc9ac43f2014612be6e7167dc7c9ff51327efa))

## [2.0.1](https://github.com/neo9/n9-node-utils/compare/2.0.0...2.0.1) (2020-10-09)


### types

* Fix types ([0b16f2d](https://github.com/neo9/n9-node-utils/commit/0b16f2d95f50e1968708ced08d7206173f609821))

# [2.0.0](https://github.com/neo9/n9-node-utils/compare/2.0.0-beta.0...2.0.0) (2020-10-09)

# [2.0.0-beta.0](https://github.com/neo9/n9-node-utils/compare/v1.10.0...2.0.0-beta.0) (2020-10-09)


### errors

* Add date and hostname to n9-errors ([1c501a1](https://github.com/neo9/n9-node-utils/commit/1c501a164d4b661dff06c44b0452d25273a72711))

### yarn

* Switch from npm to yarn ([72ce9f2](https://github.com/neo9/n9-node-utils/commit/72ce9f29d3f699da826a60547836a78ac6c5a5f3))



# [1.10.0](https://github.com/neo9/n9-node-utils/compare/v1.10.0...2.0.0-beta.0) (2019-11-14)


### errors

* Keep stack in error when stringify to JSON ([ea5b2a8](https://github.com/neo9/n9-node-utils/commit/ea5b2a8261acbd238f2b06e669347e66db4ba34e))



# [1.9.0](https://github.com/neo9/n9-node-utils/compare/v1.10.0...2.0.0-beta.0) (2019-10-10)


### error

* Handle error message from contexte when stringified ([9a6306a](https://github.com/neo9/n9-node-utils/commit/9a6306a86083996381b5adf4b75f0a4b2dbe911d))

### errors

* Fix errors toJSON function ([7546573](https://github.com/neo9/n9-node-utils/commit/7546573fd0604538b0e057c3b45992c87cfa2c55))
* Keep stack in error context ([698df48](https://github.com/neo9/n9-node-utils/commit/698df486da9c59bef1664b50fc11adcea4107885))

### json

* Fix N9Erro to JSON ([89e1f1b](https://github.com/neo9/n9-node-utils/commit/89e1f1b79f4575fd4188cf617d51ff41786d3f34))

### release

* Release 1.8.0 ([d04a963](https://github.com/neo9/n9-node-utils/commit/d04a96355aaa61551a90fae52a1ff283ec66a93d))
* Release v 1.8.1 ([22766ca](https://github.com/neo9/n9-node-utils/commit/22766ca582b62aa491c8ff3a2ae8cfe57a32a691))



## [1.7.2](https://github.com/neo9/n9-node-utils/compare/v1.10.0...2.0.0-beta.0) (2018-07-13)


### base

* Add N9Error ([4d4f798](https://github.com/neo9/n9-node-utils/commit/4d4f79854c70c5ba4a73d0e250d76fdeb7b3cc81))

### docs

* Update syntax highlight ([002549f](https://github.com/neo9/n9-node-utils/commit/002549f7bf4933cf265ecf2af40dc17339cef557))

### feat

* Add asyncObject ([fd19b7f](https://github.com/neo9/n9-node-utils/commit/fd19b7f365e19ac37b0b5d17cc718c2314c19136))
* Add cb & rename test files ([b2eeca9](https://github.com/neo9/n9-node-utils/commit/b2eeca90e25fd676407f05c54adcbb4bd7b4056d))
* Add ok(promise) ([ceffd7f](https://github.com/neo9/n9-node-utils/commit/ceffd7f95266fd6ba194dc16bdbbd48eacf293a4))
* Add waitFor & waitForEvent ([2e3519b](https://github.com/neo9/n9-node-utils/commit/2e3519b7eea93a2f44dfec1c7ceccda3571cbfc2))

### pacakge

* Upgrade n9-sonar-generate ([8bb789e](https://github.com/neo9/n9-node-utils/commit/8bb789eb67e68b5221944352ce92f7d41f2b4845))

### package

* Add n9-node-sonar-generate ([0bd3605](https://github.com/neo9/n9-node-utils/commit/0bd3605c12b70324a2b3e9b8b338aaebb7bf243c))
* Add sonar TS options ([94bd482](https://github.com/neo9/n9-node-utils/commit/94bd482531882b8bfb0edd969566e97af2e53e6e))

### readme

* fix typo ([9dc04a0](https://github.com/neo9/n9-node-utils/commit/9dc04a0ac73dfd42354f391bd0ce670383c4a8d2))

### sonar

* add sonar project to gitignore ([77ef794](https://github.com/neo9/n9-node-utils/commit/77ef7949295e29144cda4ef9da09219a8441bb58))
* delete project properties ([312307f](https://github.com/neo9/n9-node-utils/commit/312307f65f973435c6ddba6b067aa709e085c548))

### stream


test/ok.ts
import test from 'ava'

import { ok, waitFor } from '../src'

async function p(val, delay = 0) {
	await waitFor(delay)
	return val
}

async function fail(delay = 0) {
	await waitFor(delay)
	throw new Error('fail')
}

test('Returns a value on success', async (t) => {
	const foo = await ok(p('foo'))
	t.is(foo, 'foo')
})

test('Returns undefined on error', async (t) => {
	const res = await ok(fail())
	t.is(res, undefined)
})


test/index.ts
import test from 'ava'

import { N9Error } from '../src'

test('Should throw an error with a message', (t) => {
	const err = new N9Error('test')
	t.is(err.message, 'test')
	t.is(err.status, 500)
	t.deepEqual(err.context, {})
	t.true(err.stack.includes('dist/test/index.js'))
})

test('Should throw an error (status)', (t) => {
	const err = new N9Error('custom-status', 400)
	t.is(err.message, 'custom-status')
	t.is(err.status, 400)
	t.deepEqual(err.context, {})
})

test('Should throw an error (context)', (t) => {
	const err = new N9Error('custom-context', 505, { test: true })
	t.is(err.message, 'custom-context')
	t.is(err.status, 505)
	t.deepEqual(err.context, { test: true })
})


test/cb.test.ts
import ava from 'ava';

import { cb } from '../src';

function read(
	file: string,
	callback: (error?: Error | null, content?: string) => void,
): NodeJS.Timer | undefined {
	if (file === 'foo.txt') {
		return setTimeout(() => callback(null, 'contents'), 0);
	}
	setTimeout(() => callback(new Error('file not found')), 0);
}

ava.serial('Passes args and receives results', async (t) => {
	const result = await cb(read, 'foo.txt');
	t.is(result, 'contents');
});

ava.serial('Throws an error if (err) is truthy', async (t) => {
	const err = await t.throwsAsync(cb(read, 'bar.txt'));
	t.is(err.message, 'file not found');
});


src/index.ts
import { EventEmitter } from 'events';
import { NextFunction, Response } from 'express';
import * as os from 'os';
import { Transform } from 'stream';
import { promisify } from 'util';

const sleep = promisify(setTimeout);

/**
 * N9Error(message [, status] [, context])
 *
 *  @param status The HTTP Status of the error. i.e. : 404 → Not Found
 *  @param context Should be an object helping to understand the error
 */
export class N9Error extends Error {
	public readonly date: Date;
	public readonly hostname: string;
	public readonly message: string;
	public readonly status: number;
	public context: any;

	constructor(message: string, status?: number, context?: Record<string, any>) {
		super(message);
		this.date = new Date();

		// we check if an error has been passed in the context
		// if so we create an intermediary object
		// to allow JSON.stringify to work correctly render the error's message
		if (context) {
			for (const key of Object.keys(context)) {
				if (context[key] instanceof Error) {
					context[key] = {
						...context[key],
						message: context[key].message,
						name: context[key].name,
						stack: context[key].stack,
					};
				}
			}
		}

		this.message = message;
		this.status = status || 500;
		this.context = context || {};
		this.hostname = os.hostname();
		Error.captureStackTrace(this, N9Error);
	}

	/**
	 *
	 */
	public toJSON?(): object {
		return {
			name: this.name,
			message: this.message,
			status: this.status,
			context: this.context,
			stack: this.stack,
			date: this.date,
			hostname: this.hostname,
		};
	}
}

/*
 ** ok(promise: Promise): <Promise>
 */
export async function ok(promise: Promise<any>): Promise<any> {
	try {
		return await promise;
	} catch (err) {
		return;
	}
}

/*
 ** cb(fn: Function, args: ...any): <Promise>
 */
export async function cb<T = any>(fn: (...args: any[]) => any, ...args: any[]): Promise<T> {
	return new Promise<T>((resolve, reject) => {
		fn(...args, (err: Error, result: T) => {
			if (err) return reject(err);
			resolve(result);
		});
	});
}

/*
 ** waitFor(ms)
 */
export async function waitFor(ms?: number): Promise<any> {
	await sleep(ms || 0);

	return;
}

/*
 ** waitForEvent(emmiter, eventName)
 */
export async function waitForEvent(emmiter: EventEmitter, eventName: string): Promise<any> {
	return new Promise((resolve) => {
		emmiter.once(eventName, (...args) => resolve([...args]));
	});
}

/*
 ** asyncObject(obj): Promise<Object>
 */
export async function asyncObject(
	obj: Record<string, Promise<any> | any> = {},
): Promise<Record<string, any>> {
	const keys = Object.keys(obj).filter(
		(key: string) => obj[key] && typeof obj[key].then === 'function',
	);
	const promises = keys.map((key) => obj[key]);
	const results = await Promise.all(promises);
	const container = { ...obj };
	results.forEach((result, index) => {
		const key = keys[index];
		container[key] = result;
	});
	return container;
}

export interface N9JSONStreamOptionsBase<M = any> {
	total: number;
	limit?: number;
	offset?: number;
	count?: number;
	metaData?: M;
}

export interface N9JSONStreamOptions<M = any> extends N9JSONStreamOptionsBase<M> {
	res?: Response;
	key?: string;
}

/**
 * N9JSONStreamResponse
 */
export interface N9JSONStreamResponse<T, M = any> {
	items: T[];
	count: number;
	total: number;
	limit?: number;
	offset?: number;
	metaData?: M;
}

/**
 * N9JSONStream(baseObject)
 *
 */
export class N9JSONStream<T = object, M = any> extends Transform {
	private first: boolean;
	private readonly base: N9JSONStreamOptionsBase<M>;
	private readonly key: string;

	/**
	 * Usage example :
	 * ```
	 * const cursor = await this.service.find(filter, page, size);
	 * return cursor.pipe(
	 * new N9JSONStream({
	 * res, // express response
	 * total: await cursor.count(),
	 * }),
	 * );
	 *
	 * ```
	 *
	 * @param options should give at least Express Response to stream in, and the total number of elements
	 */
	constructor(options: N9JSONStreamOptions<M>) {
		super({
			readableObjectMode: true,
			writableObjectMode: true,
		});

		if (typeof options.total === 'undefined') {
			throw new N9Error(`'total' property is required to N9JSONStream class`);
		}

		if (options.res) {
			options.res.setHeader('Content-Type', 'application/json; charset=utf-8');
		}

		this.first = true;
		this.base = {
			total: options.total,
			limit: options.limit,
			offset: options.offset,
			metaData: options.metaData,
			count: 0,
		};
		this.key = options.key || 'items';
		this.push(`{"${this.key}":[`);
	}

	public _transform(item: T, encoding: string, next: NextFunction): void {
		if (!this.first) this.push(',');
		else this.first = false;

		this.push(JSON.stringify(item));
		this.base.count += 1;

		return next();
	}

	public _flush(next: NextFunction): void {
		const keys: (keyof N9JSONStreamOptionsBase<M>)[] = [
			'limit',
			'offset',
			'total',
			'count',
			'metaData',
		];

		const keysWithValue = keys.filter((key) => typeof this.base[key] !== 'undefined');

		this.push('],');
		keysWithValue.forEach((key, i) => {
			const value = this.base[key];
			if (typeof value === 'object') {
				this.push(`"${key}": ${JSON.stringify(value)}`);
			} else {
				// eslint-disable-next-line @typescript-eslint/no-base-to-string
				this.push(`"${key}": ${value}`);
			}
			if (i < keysWithValue.length - 1) this.push(',');
		});
		this.push('}');

		return next();
	}
}
